crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -0,0 +1,764 @@
+                return Ok(())
+            }
+        };
+
+        Ok(())
+    }
+
+    fn try_reconfiguring(
+        &mut self,
+        observed_cluster_state: &ObservedClusterState,
+        chain_builder: &mut ChainBuilder,
+    ) -> Result<(), anyhow::Error> {
+        match self {
+            LogsState::Sealed {
+                seal_lsn,
+                configuration,
+                ..
+            } => {
+                if let Some(loglet_configuration) =
+                    configuration.try_reconfiguring(observed_cluster_state)
+                {
+                    let segment_index = chain_builder.append_segment(
+                        *seal_lsn,
+                        loglet_configuration.as_provider(),
+                        loglet_configuration.to_loglet_params()?,
+                    )?;
+
+                    *self = LogsState::Running {
+                        segment_index,
+                        configuration: Some(loglet_configuration),
+                    };
+                }
+            }
+            // nothing to do :-)
+            LogsState::Provisioning { .. }
+            | LogsState::Sealing { .. }
+            | LogsState::Running { .. } => return Ok(()),
+        };
+
+        Ok(())
+    }
+
+    fn requires_reconfiguration(&self, observed_cluster_state: &ObservedClusterState) -> bool {
+        match self {
+            LogsState::Provisioning { .. }
+            | LogsState::Sealing { .. }
+            | LogsState::Sealed { .. } => false,
+            LogsState::Running { configuration, .. } => configuration
+                .as_ref()
+                .expect("configuration must be present")
+                .requires_reconfiguration(observed_cluster_state),
+        }
+    }
+
+    fn current_segment(&self) -> Option<SegmentIndex> {
+        match self {
+            LogsState::Provisioning { .. } => None,
+            LogsState::Running { segment_index, .. } => Some(*segment_index),
+            LogsState::Sealing { segment_index, .. } => Some(*segment_index),
+            LogsState::Sealed { segment_index, .. } => Some(*segment_index),
+        }
+    }
+}
+
+fn try_provisioning(
+    provider_kind: ProviderKind,
+    observed_cluster_state: &ObservedClusterState,
+) -> Option<LogletConfiguration> {
+    match provider_kind {
+        ProviderKind::Local => {
+            let log_id = thread_rng().next_u64();
+            Some(LogletConfiguration::Local(log_id))
+        }
+        #[cfg(any(test, feature = "memory-loglet"))]
+        ProviderKind::InMemory => {
+            let log_id = thread_rng().next_u64();
+            Some(LogletConfiguration::Memory(log_id))
+        }
+        ProviderKind::Replicated => {
+            find_new_replicated_loglet_configuration(observed_cluster_state, None)
+                .map(LogletConfiguration::Replicated)
+        }
+    }
+}
+
+fn find_new_replicated_loglet_configuration(
```

Comment:
This name confused me, maybe `find_` -> `build_`?

Reply:
💯 Will give it a better name.